### PR TITLE
gc_worker: fix serveral bugs of green gc

### LINKED ISSUE #16306: The retry of physical resolve lock is useless
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 3. What did you see instead?
1. GC worker doesn't re-register lock observers of dirty stores which causes these stores are always dirty.
2. GC worker doesn't register lock observers of newly added stores which causes these stores are always dirty.
3. Can't remove lock observers of cleaned stores.
4. If there are dirty stores after 3 round physical resolve lock, GC worker won't fall back to legacy procedure which breaks the data safety.

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
4.0

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?

Issue Number: close #16306 

Problem Summary:
1. GC worker doesn't re-register lock observers of dirty stores which causes these stores are always dirty.
2. GC worker doesn't register lock observers of newly added stores which causes these stores are always dirty.
3. Can't remove lock observers of cleaned stores.
4. If there are dirty stores after 3 round physical resolve lock, GC worker won't fall back to legacy procedure which breaks the data safety.

### What is changed and how it works?
What's Changed:
It's simple. Please refer to the codes.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

### Release note 
Fix the issue that green GC may leave unresolved locks if there are dirty stores.

### CODE REVIEW

File: store/tikv/gcworker/gc_worker.go [hunk 1]
```diff
@@ -1066,20 +1073,31 @@ func (w *GCWorker) resolveLocksPhysical(ctx context.Context, safePoint uint64) e
...
+		for store := range stores {
+			if _, ok := checkedStores[store]; ok {
+				// The store is resolved and checked.
+				if _, ok := resolvedStores[store]; ok {
+					delete(stores, store)
+				}
+				// The store is checked and has been resolved before.
+				if _, ok := dirtyStores[store]; !ok {
+					delete(stores, store)
+				}
```
Comment (MyonKeminta) [inline store/tikv/gcworker/gc_worker.go:1085]: resolved stores may become dirty again.

Comment (youjiali1995) [reply store/tikv/gcworker/gc_worker.go:1085]: Yes. It will be scanned and resolved again.

File: store/tikv/gcworker/gc_worker.go [hunk 2]
```diff
@@ -1066,20 +1073,31 @@ func (w *GCWorker) resolveLocksPhysical(ctx context.Context, safePoint uint64) e
...
+		dirtyStores = stores
 
 		// If there are still dirty stores, continue the loop to clean them again.
 		// Only dirty stores will be scanned in the next loop.
-		if len(stores) == 0 {
+		if len(dirtyStores) == 0 {
 			break
 		}
+		// Clear all collected locks of dirty stores to prevent it from being dirty again.
+		w.removeLockObservers(ctx, safePoint, dirtyStores)
```
Comment (youjiali1995) [inline store/tikv/gcworker/gc_worker.go:1096]: I found we can't do it. If the store recevies duplicated `removeLockObserver` and `registerLockObserver` between L1061 and L1071. The store is cleaned but the lock observer drops some locks.

Comment (youjiali1995) [reply store/tikv/gcworker/gc_worker.go:1096]: Maybe we can't retry here.

Comment (sticnarf) [reply store/tikv/gcworker/gc_worker.go:1096]: Maybe we shouldn't do "remove and re-register" on stores which fail to be checked. Instead, can we just register for those newly added stores?

Comment (youjiali1995) [reply store/tikv/gcworker/gc_worker.go:1096]: Yes. If dirty stores are newly added, we can retry. If a store is dirty due to too many collected locks, we don't retry and falls back to legacy mode.

**Pull request merged**